nemo_aligner/utils/parallel_state.py

```diff
@@ -14,18 +14,87 @@
 
 """Wrapper around mcore parallel state to handle cases of resharding"""
 
+from contextlib import contextmanager
+
 from megatron.core import parallel_state as mcore_parallel_state
 
 from nemo.collections.nlp.modules.common.text_generation_utils import (
     get_model_parallel_src_rank as nemo_get_model_parallel_src_rank,
 )
 
+_TRT_LLM_RESHARD = False
+
+
+def enable_trt_llm_reshard_calls():
+    global _TRT_LLM_RESHARD
+    _TRT_LLM_RESHARD = True
+
+
+def disable_trt_llm_reshard_calls():
+    global _TRT_LLM_RESHARD
+    _TRT_LLM_RESHARD = False
+
+
+def is_trt_llm_reshard():
+    return _TRT_LLM_RESHARD
+
 
 def get_model_parallel_src_rank():
-    src_rank = nemo_get_model_parallel_src_rank()
+    src_rank = (
+        mcore_parallel_state.get_tensor_model_parallel_src_rank()
+        if is_trt_llm_reshard()
+        else nemo_get_model_parallel_src_rank()
+    )
 
     return src_rank
 
 
+def get_model_parallel_group():
+    group = (
+        mcore_parallel_state.get_tensor_model_parallel_group()
+        if is_trt_llm_reshard()
+        else mcore_parallel_state.get_model_parallel_group()
+    )
+    return group
+
+
+def get_data_parallel_world_size():
+    data_parallel_size = mcore_parallel_state.get_data_parallel_world_size()
+
+    return (
+        data_parallel_size * mcore_parallel_state.get_pipeline_model_parallel_world_size()
+        if is_trt_llm_reshard()
+        else data_parallel_size
+    )
+
+
+def get_data_parallel_rank():
+    data_parallel_rank = mcore_parallel_state.get_data_parallel_rank()
+
+    if is_trt_llm_reshard():
+        data_parallel_rank = data_parallel_rank + (
+            mcore_parallel_state.get_data_parallel_world_size()
+            * mcore_parallel_state.get_pipeline_model_parallel_rank()
+        )
+
+    return data_parallel_rank
+
+
+def get_pipeline_model_parallel_world_size():
+    return 1 if is_trt_llm_reshard() else mcore_parallel_state.get_pipeline_model_parallel_world_size()
+
+
+@contextmanager
+def trt_llm_reshard_region():
+    """mutates global state so distributed call are aware of TRT-LLM resharding
+        from PP to TP only
+    """
+    try:
+        enable_trt_llm_reshard_calls()
+        yield
+    finally:
+        disable_trt_llm_reshard_calls()
+
+
 def __getattr__(name):
     return getattr(mcore_parallel_state, name)
```

Comment:
Are you confident this is safe?

(the alternative would be to import from `mcore_parallel_state` only the functions we actually need in NeMo-Aligner, to be sure they're all vetted as "reshard-compatible")

Reply:
> Are you confident this is safe?

no :(, I could import all the stuff from mcore_parallel_state but I worry this will be a lot of functions see [here](https://github.com/NVIDIA/Megatron-LM/blob/ef85bc94fc744aa5d398d12140f808023afbf78d/megatron/core/parallel_state.py#L4). 

I've added a warning when reshard is on but we defaulted back to mcore 9c165a066fe1e237954a7fcc61dda2fe1d563529

Reply:
Yeah definitely don't import everything. But I suspect we only use a pretty limited set, especially when constrained by `is_trt_llm_reshard()` => could we just crash instead of a warning? (with a clear error message that says to either import this function in `parellel_state.py` if it's already "reshard-compatible", or implement a reshard-compatible version)

Reply:
hmm I worry that this might be annoying in the long run. because anytime anyone needs to implement something new(say a new type of parallelism) they will be forced to create a reshard compatible one even if they do not use it or overwrite it with a dummy function that calls into mcore_parallel_state

instead I currently think of reshard as an opt in feature -> if you enable reshard then you are responsible for implementing the resharding ops in parallel state.

I do agree with your statement though if aligner decides to support reshard more broadly but I'm not sure it's worth the complexity

Reply:
> because anytime anyone needs to implement something new(say a new type of parallelism) they will be forced to create a reshard compatible one even if they do not use it or overwrite it with a dummy function that calls into mcore_parallel_state

Actually that's not what I meant in my previous message: if you replace the warning with a crash, it won't trigger unless someone uses it with reshard (since it's guarded by a `if is_trt_llm_reshard()`). And if the correct approach is just to re-use `mcore_parallel_state` (even with reshard), then you don't need to write a dummy function, just to import it, so it's quite trivial.

It's not a huge deal but I see the overhead as minor, while the potential benefits (of saving some potentially meaningful debugging time to understand why something doesn't work well when we enable TRT-LLM) seem more significant to me.

Reply:
ah sorry I misunderstood your concern. agree with the above.

4918eafcda3feb0128e5f8b790ed700ce42436c5